dsspence/mattermost-plugin-github
# [MM-409] Add subscription features for project issues and project pull requests

### Summary

This PR is an initial attempt to add a new subscription feature for issues and pull requests assigned to a project https://github.com/mattermost/mattermost-plugin-github/issues/409 .

Since the issue was first proposed, GitHub has introduced a breaking change to their projects implementation creating both a projects (classic) and a projects v2 split. I had to first upgrade the google/go-github dependency to support the newer version of projects https://github.com/mattermost/mattermost-plugin-github/pull/771. When I tried to create an older project resource through the REST API it rejected the request.

Projects V2 are able to be created at the organizational level, user level, and repository level.  This brings into question how much scoping a subscription to a particular repository is to be implemented if projects exist outside of repo boundaries.

I began by trying to re-use the existing `issues` and `pulls` features in the same way as labels are implemented. I quickly decided that this would mean overloading the usage of those features in ways that would make the codebase even more complex such as around duplicate webhook events.  I ultimately decided to segment these into new features titled `project_issues` and `project_pulls`.  There is a new draft issue item as well for projects that doesn't map cleanly to the supported features.

At subscription time this code will use the GraphQL API to find all projects (org and repo level) belonging to the organization and validate that the project title matches one that the user can access (projects can be public or private). This required adding a new oauth scope to the plugin `read:project`, which it didn't have previously.
  
Finally a new webhook event handler was needed to receive the Project V2 Item event, mapping it to an existing project subscription, and building a post action template to send to the Mattermost server.

### Testing Instructions
1. Verify or update your GitHub webhook to include the project v2 items
![image](https://github.com/dsspence/mattermost-plugin-github/assets/111698/587be93a-bb5d-4e90-b1e6-48f93b987f94)
2. Deploy this plugin to your local running Mattermost development server using
```sh
make deploy
```
3. From a channel or private message create a subscription to an organization with an active project
```sh
/github subscriptions add dsspence-org --features project_issues,project:"org project"
```
4. Create a new issue and assign the project used in the previous step from the sidebar before submitting.
5. Observe the new project issue created post in the subscribed Mattermost channel/DM.

### Remaining Steps
- Write automated tests
- Expand GraphQL query to include projects owned by users not just organizations and support pagination beyond first 100 results
- Edge cases such as non-unique project names, trim whitespace, reject empty values ...
- Introduce GraphQL query to lookup issue and pull_requests associated in the Project V2 Item webhook event to provide more detailed post info and/or repo scoping.
- Add pull request template and draft issue support
- Support all of the Project V2 Item event actions beyond just creation (archiving, deleted, restored) and custom templates
- Verify permissions are what we expect in different scenarios

### Open Questions / Observations
- Duplicative default issue subscription is still being shown alongside project issue item subscription
- There was a server setup step around enabling CORS that I would probably add
- A plugin deploy will fail silently if there is a template error, this could use a better error message or additional build step.
- I noticed subscribing with a label only ever uses the first one provided and doesn't support multiple labels in practice. Maybe this should be updated to clarify or support added for checking for all labels in a subscription.
- When attempting to run e2e tests, I think there is an issue with the latest version of Playwright that requires using an older version. I had trouble with the node postinstall rollup command failing for the Mattermost server `webapp/platform/components` package. I had to install dependencies from the child directories because it was failing from the parent folder.
- Node 16 is EOL, are there blockers on moving up to Node 18?

### References
[Organization GraphQL API](https://docs.github.com/en/graphql/reference/queries#organization)
[Projects V2 Item Webhook Event](https://docs.github.com/en/webhooks/webhook-events-and-payloads#projects_v2_item)
[Projects V2 Overview](https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects)

https://github.com/dsspence/mattermost-plugin-github/assets/111698/22450c61-320e-437d-a4b1-8be7a8f98352
